src/main/java/org/wildfly/security/auth/realm/ldap/AttributeMapping.java

```diff
@@ -51,76 +104,116 @@ public static AttributeMapping from(String ldapName) {
+        builder.name = builder.ldapName;
+        return builder;
     }
 
     /**
-     * <p>The behavior is exactly the same as {@link #fromFilter(String, String, String)}, except that it uses the
-     * same name of the context defined in {@link org.wildfly.security.auth.realm.ldap.LdapSecurityRealmBuilder.IdentityMappingBuilder#setSearchDn(String)}.
+     * <p>The behavior is exactly the same as {@link #fromFilter(String, String)}, except that instead
+     * of attribute value is DN of found entries used. As the search DN is used search DN from identity mapping.
      *
      * @param filter the filter that is going to be used to search for entries and obtain values for this attribute
-     * @param ldapName the name of the attribute in LDAP from where the values are obtained
      * @return this builder
      */
-    public static AttributeMapping fromFilter(String filter, String ldapName) {
+    public static Builder fromFilterDn(String filter) {
         Assert.checkNotNullParam("filter", filter);
-        Assert.checkNotNullParam("ldapName", ldapName);
-        return new AttributeMapping(null, filter, ldapName);
-    }
-
-    AttributeMapping(String ldapName) {
-        this(null, null, ldapName);
-    }
-
-    AttributeMapping(String searchDn, String filter, String ldapName) {
-        Assert.checkNotNullParam("ldapName", ldapName);
-        this.searchDn = searchDn;
-        this.filter = filter;
-        this.ldapName = ldapName.toUpperCase();
+        Builder builder = new Builder();
+        builder.filter = filter;
+        builder.name = "dn";
+        return builder;
     }
 
-    public AttributeMapping asRdn(String rdn) {
-        Assert.checkNotNullParam("rdn", rdn);
-        this.rdn = rdn;
-        return this;
-    }
+    public static class Builder {
+
+        private String ldapName;
+        private String searchDn;
+        private boolean recursiveSearch = true;
+        private String filter;
+        private String name;
+        private String rdn;
+        private int recursiveDepth;
+
+        /**
+         * Set type of RDN, whose value will be used as identity attribute value.
+         * Use in case the attribute value is in DN form or when DN of entry is used.
+         *
+         * @param rdn the name of type of RDN
+         * @return this builder
+         */
+        public Builder extractRdn(String rdn) {
+            Assert.checkNotNullParam("rdn", rdn);
+            this.rdn = rdn;
+            return this;
+        }
 
-    public AttributeMapping to(String name) {
-        Assert.checkNotNullParam("to", name);
-        this.name = name;
-        return this;
-    }
+        /**
+         * Set name of identity attribute to which will be mapping done.
+         *
+         * @param name the name of identity attribute (not LDAP attribute)
+         * @return this builder
+         */
+        public Builder to(String name) {
+            Assert.checkNotNullParam("to", name);
```

Comment:
This was already a problem with this in the old code, but the parameter name strings and values are mismatched here.